File: examples/high_quality_solution/espressif/components/esp_board_manager/devices/dev_lcd_touch_i2c/dev_lcd_touch_i2c.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include "esp_log.h"

#include "esp_check.h"
#include "driver/gpio.h"
#include "driver/i2c_master.h"
#include "esp_lcd_panel_io.h"
#include "esp_lcd_touch.h"

#include "dev_lcd_touch_i2c.h"
#include "esp_board_periph.h"
#include "esp_board_device.h"

static const char *TAG = "DEV_LCD_TOUCH_I2C";

// Touch driver factory function declarations
extern esp_err_t lcd_touch_factory_entry_t(const esp_lcd_panel_io_handle_t io, const esp_lcd_touch_config_t *config, esp_lcd_touch_handle_t *tp);

int dev_lcd_touch_i2c_init(void *cfg, int cfg_size, void **device_handle)
{
    dev_lcd_touch_i2c_config_t *touch_cfg = (dev_lcd_touch_i2c_config_t *)cfg;

    dev_lcd_touch_i2c_handles_t *touch_handles = (dev_lcd_touch_i2c_handles_t *)calloc(1, sizeof(dev_lcd_touch_i2c_handles_t));
    if (touch_handles == NULL) {
        ESP_LOGE(TAG, "Failed to allocate touch handles");
        return -1;
    }
    ESP_LOGI(TAG, "Initializing LCD touch: %s, chip: %s, addr: 0x%02" PRIx32, touch_cfg->name, touch_cfg->chip, touch_cfg->io_i2c_config.dev_addr);

    // Get I2C peripheral handle
    void *i2c_bus_handle = NULL;
    int ret = 0;
    if (touch_cfg->i2c_name && strlen(touch_cfg->i2c_name) > 0) {
        ret = esp_board_periph_ref_handle(touch_cfg->i2c_name, &i2c_bus_handle);
        if (ret != 0) {
            ESP_LOGE(TAG, "Failed to get I2C peripheral handle: %d", ret);
            free(touch_handles);
            return -1;
        }
    } else {
        ESP_LOGE(TAG, "No I2C name configured for LCD touch: %s", touch_cfg->name);
        free(touch_handles);
        return -1;
    }

    esp_lcd_panel_io_i2c_config_t io_i2c_config = {};
    memcpy(&io_i2c_config, &touch_cfg->io_i2c_config, sizeof(io_i2c_config));
    io_i2c_config.dev_addr = 0x00;
    for (size_t i = 0; i < sizeof(touch_cfg->i2c_addr) / sizeof(touch_cfg->i2c_addr[0]); i++) {
        if (touch_cfg->i2c_addr[i] > 0) {
            ESP_LOGD(TAG, "I2C handle: %p, dev_addr[%zu]: 0x%02x, scl_speed_hz: %" PRIu32,
                i2c_bus_handle, i, touch_cfg->i2c_addr[i], touch_cfg->io_i2c_config.scl_speed_hz);
            ret = i2c_master_probe(i2c_bus_handle, touch_cfg->i2c_addr[i], 200/portTICK_PERIOD_MS);
            if (ret == ESP_OK) {
                io_i2c_config.dev_addr = touch_cfg->i2c_addr[i];
                break;
            }
        }
    }
    ret = esp_lcd_new_panel_io_i2c_v2((i2c_master_bus_handle_t)i2c_bus_handle, &io_i2c_config, &touch_handles->io_handle);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to create LCD panel IO for touch, i2c addr: %" PRIx32, io_i2c_config.dev_addr);
        free(touch_handles);
        return -1;
    }

    ESP_LOGD(TAG, "touch_config: x_max=%d, y_max=%d, rst_gpio_num=%d, int_gpio_num=%d, levels.reset=%d, levels.interrupt=%d, flags.swap_xy=%d, flags.mirror_x=%d, flags.mirror_y=%d",
             touch_cfg->touch_config.x_max,
             touch_cfg->touch_config.y_max,
             touch_cfg->touch_config.rst_gpio_num,
             touch_cfg->touch_config.int_gpio_num,
             touch_cfg->touch_config.levels.reset,
             touch_cfg->touch_config.levels.interrupt,
             touch_cfg->touch_config.flags.swap_xy,
             touch_cfg->touch_config.flags.mirror_x,
             touch_cfg->touch_config.flags.mirror_y);

    ret = lcd_touch_factory_entry_t(touch_handles->io_handle, &touch_cfg->touch_config, &touch_handles->touch_handle);
    if (ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to create CST816S touch driver: %s", esp_err_to_name(ret));
        esp_lcd_panel_io_del(touch_handles->io_handle);
        free(touch_handles);
        return -1;
    }

    ESP_LOGI(TAG, "Successfully initialized: %s, touch: %p, io: %p", touch_cfg->name, touch_handles->touch_handle, touch_handles->io_handle);
    *device_handle = touch_handles;
    return 0;
}

int dev_lcd_touch_i2c_deinit(void *device_handle)
{
    dev_lcd_touch_i2c_handles_t *touch_handles = (dev_lcd_touch_i2c_handles_t *)device_handle;

    if (touch_handles->touch_handle) {
        esp_lcd_touch_del(touch_handles->touch_handle);
        touch_handles->touch_handle = NULL;
    }

    if (touch_handles->io_handle) {
        esp_lcd_panel_io_del(touch_handles->io_handle);
        touch_handles->io_handle = NULL;
    }

    const char *name = NULL;
    const esp_board_device_handle_t *device_handle_struct = esp_board_device_find_by_handle(device_handle);
    if (device_handle_struct) {
        name = device_handle_struct->name;
    }
    dev_lcd_touch_i2c_config_t *cfg = NULL;
    esp_board_device_get_config(name, (void **)&cfg);
    if (cfg) {
        esp_board_periph_unref_handle(cfg->i2c_name);
    }

    free(touch_handles);
    return 0;
}

File: examples/high_quality_solution/espressif/components/esp_board_manager/devices/dev_lcd_touch_i2c/dev_lcd_touch_i2c.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# LCD touch I2C device config parser
VERSION = 'v1.0.0'

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

def get_includes() -> list:
    """Return list of required include headers for LCD touch I2C device"""
    return [
        'dev_lcd_touch_i2c.h'
    ]

def parse(name: str, config: dict, peripherals_dict=None) -> dict:
    """Parse LCD touch I2C device configuration from YAML to C structure"""

    # Parse the device name - use name directly for C naming
    c_name = name.replace('-', '_')  # Convert hyphens to underscores for C naming

    # Get the device config and peripherals
    device_config = config.get('config', {})
    peripherals = config.get('peripherals', [])

    # Get chip name from device level
    chip_name = config.get('chip', 'unknown')

    # Get nested configurations
    io_i2c_config = device_config.get('io_i2c_config', {})
    touch_config = device_config.get('touch_config', {})

    # Initialize default configurations
    i2c_cfg = {'name': '', 'port': 0, 'address': 0x5D, 'frequency': 400000}

    # Process I2C peripheral from io_i2c_config.peripherals
    io_i2c_peripherals = io_i2c_config.get('peripherals', [])
    for periph in io_i2c_peripherals:
        periph_name = periph.get('name', '')
        if periph_name.startswith('i2c'):
            # Check if peripheral exists in peripherals_dict if provided
            if peripherals_dict is not None and periph_name not in peripherals_dict:
                raise ValueError(f"Device '{name}' references undefined peripheral '{periph_name}'")
            # This is I2C configuration
            address = periph.get('address', 0x5D)  # Keep hex format
            # Get port from peripheral config if available
            port = 0
            if peripherals_dict and periph_name in peripherals_dict:
                peripheral_config = peripherals_dict[periph_name]
                port = peripheral_config.config.get('port', 0)
            i2c_cfg = {
                'name': periph_name,  # Use original name from YAML
                'port': port,
                'address': address,  # No need to convert, keep as hex
                'frequency': int(periph.get('frequency', 400000))
            }

    # Parse I2C interface settings
    dev_addr_raw = io_i2c_config.get('dev_addr', 0x5a)

    # Handle dev_addr configuration: support both single value and array
    if isinstance(dev_addr_raw, list):
        # Array format: [primary_addr, secondary_addr]
        i2c_addr_primary = int(dev_addr_raw[0]) if len(dev_addr_raw) > 0 else 0x5a
        i2c_addr_secondary = int(dev_addr_raw[1]) if len(dev_addr_raw) > 1 else 0
    else:
        # Single value format: use as primary address, secondary = 0
        i2c_addr_primary = int(dev_addr_raw)
        i2c_addr_secondary = 0

    # Keep dev_addr for backward compatibility (use primary address)
    dev_addr = i2c_addr_primary

    control_phase_bytes = int(io_i2c_config.get('control_phase_bytes', 1))
    dc_bit_offset = int(io_i2c_config.get('dc_bit_offset', 0))
    lcd_cmd_bits = int(io_i2c_config.get('lcd_cmd_bits', 8))
    lcd_param_bits = int(io_i2c_config.get('lcd_param_bits', 0))
    scl_speed_hz = int(io_i2c_config.get('scl_speed_hz', 100000))

    # Parse I2C flags from io_i2c_config.flags
    io_i2c_flags = io_i2c_config.get('flags', {})
    dc_low_on_data = bool(io_i2c_flags.get('dc_low_on_data', False))
    disable_control_phase = bool(io_i2c_flags.get('disable_control_phase', True))

    # Parse touch configuration
    x_max = int(touch_config.get('x_max', 320))
    y_max = int(touch_config.get('y_max', 240))
    rst_gpio_num = int(touch_config.get('rst_gpio_num', -1))
    int_gpio_num = int(touch_config.get('int_gpio_num', -1))

    # Parse touch levels
    levels = touch_config.get('levels', {})
    reset_level = int(levels.get('reset', 0))
    interrupt_level = int(levels.get('interrupt', 0))

    # Parse touch flags
    touch_flags = touch_config.get('flags', {})
    swap_xy = bool(touch_flags.get('swap_xy', False))
    mirror_x = bool(touch_flags.get('mirror_x', False))
    mirror_y = bool(touch_flags.get('mirror_y', False))

    result = {
        'struct_type': 'dev_lcd_touch_i2c_config_t',
        'struct_var': f'{c_name}_cfg',
        'struct_init': {
            'name': c_name,
            'chip': chip_name,
            'type': str(config.get('type', 'lcd_touch_i2c')),
            'i2c_name': i2c_cfg['name'],
            'i2c_addr': [f'0x{i2c_addr_primary:02x}', f'0x{i2c_addr_secondary:02x}'],

            # IO I2C Configuration (esp_lcd_panel_io_i2c_config_t)
            'io_i2c_config': {
                'dev_addr': dev_addr,
                'control_phase_bytes': control_phase_bytes,
                'dc_bit_offset': dc_bit_offset,
                'lcd_cmd_bits': lcd_cmd_bits,
                'lcd_param_bits': lcd_param_bits,
                'scl_speed_hz': scl_speed_hz,
                'flags': {
                    'dc_low_on_data': dc_low_on_data,
                    'disable_control_phase': disable_control_phase
                }
            },

            # Touch Configuration (esp_lcd_touch_config_t)
            'touch_config': {
                'x_max': x_max,
                'y_max': y_max,
                'rst_gpio_num': rst_gpio_num,
                'int_gpio_num': int_gpio_num,
                'levels': {
                    'reset': reset_level,
                    'interrupt': interrupt_level
                },
                'flags': {
                    'swap_xy': swap_xy,
                    'mirror_x': mirror_x,
                    'mirror_y': mirror_y
                },
                'process_coordinates': None,
                'interrupt_callback': None,
                'user_data': None,
                'driver_data': None
            }
        }
    }

    return result
